# TDI Capstone - Final Report

### Abstract:
Phase III clincial trials have a massive effect on the market captalizatizations of pharmaceutical companies. A successful phase III trial for a novel drug allows a company to begin marketing drugs to a previously unserved clinical indication, yielding large revenue streams. In order for a drug to pass it must pass through roughly 10 years of Research and be tested on thousands (or tens of thousands) of patients. 

The long time scales of trials and the sheer numbers of people involved (patients, research scientists, clinicans, research coordinators, goverment regulators...), means that priveleged information regarding trial status has an unusally high level of exposure, compared to other high tech industries. [Recent research](https://academic.oup.com/jnci/article/103/20/1507/904625/Company-Stock-Prices-Before-and-After-Public) suggests that this information exposure can cause detectable movement in this public stock markets. 

If leaks of priveleged information can affect the valuations of pharmaceutical companies, can the movements of these valutaions be identified using Machine Learning and used to inform smarter trading decisions? 

### Gathering Data:

To start with this problem, we need to select a range of target companies (in the pharma sector), find daily closing prices of thier stocks, and get the dates of thier relevant approval announcements. With this information, I'll attempt to extract features and train a machine learning model. 

Data Sources: 
* [fdaTracker.com's free PDUFA Calendar](https://www.fdatracker.com/fda-calendar/)
* [Biopharm Catalyst's upcoming PDUFA Calendar](https://www.biopharmcatalyst.com/calendars/fda-calendar)
* [AlphaVantage's Stock Price API](https://www.alphavantage.co/)

###### First:
First, lets get the historical PDUFA (FDA announcement) Dates:

In [378]:
from urllib2 import urlopen
import ics
import re
from datetime import datetime, timedelta
from alpha_vantage.timeseries import TimeSeries
from tqdm import tqdm_notebook
import numpy as np
import pandas as pd
import dill
from random import randint

In [2]:
tickerRe = re.compile(r"\A[A-Z]{3,4}\W")
today = datetime.today()

FdaUrl = "https://calendar.google.com/calendar/ical/5dso8589486irtj53sdkr4h6ek%40group.calendar.google.com/public/basic.ics"
FdaCal = ics.Calendar(urlopen(FdaUrl).read().decode('iso-8859-1'))
FdaCal

<Calendar with 552 events>

In [3]:
past_pdufa_syms = set()
for event in FdaCal.events:
    matches = re.findall(tickerRe, event.name)
    if len(matches) >=1:
        eComp = str(matches[0]).strip().strip(".")
        past_pdufa_syms.add(eComp)

In [4]:
print past_pdufa_syms

set(['ENTA', 'AAAP', 'PAR', 'KITE', 'BSTC', 'DNDN', 'ELTP', 'ZSPH', 'GSK', 'CLVS', 'ADLR', 'XENT', 'IRWD', 'VNDA', 'RPTP', 'ACUR', 'DEPO', 'ARIA', 'CHMA', 'VRTX', 'OREX', 'THRX', 'HPTX', 'PCRX', 'ENDP', 'NAVB', 'DRRX', 'OTIC', 'EXEL', 'FLXN', 'ZGEN', 'DSCO', 'SRPT', 'PSDV', 'BCRX', 'MRK', 'ADMP', 'ADMS', 'ISTA', 'AMRN', 'BMY', 'ITMN', 'CEMP', 'KMPH', 'RLYP', 'ANAC', 'SUPN', 'JNJ', 'AERI', 'SNY', 'VION', 'AMAG', 'HZNP', 'REGN', 'SOMX', 'COLL', 'CRTX', 'LPCN', 'VCEL', 'GTXI', 'CTIC', 'LLY', 'AFFY', 'CYPB', 'HGSI', 'OCUL', 'AEGR', 'HOLX', 'OSI', 'DVAX', 'TEVA', 'SGEN', 'TTNP', 'ACOR', 'RPRX', 'EGRX', 'JAZZ', 'AUXL', 'QCOR', 'FLML', 'NEOS', 'AGRX', 'RARE', 'RDUS', 'NGSX', 'RMTI', 'NPSP', 'CLDA', 'DYAX', 'ALTH', 'ARLZ', 'VVUS', 'LXRX', 'NVO', 'MELA', 'ANX', 'SCMP', 'POZN', 'HRTX', 'ACAD', 'AGIO', 'TSRO', 'RGEN', 'FURX', 'SPPI', 'AMGN', 'ONXX', 'BMTI', 'BPAX', 'REPH', 'DOR', 'NEW', 'WCRX', 'ASTX', 'KTOV', 'ISIS', 'AVEO', 'BIOD', 'GNBT', 'VRX', 'IGXT', 'LGND', 'MDVN', 'AMLN', 'OMER', 'KMDA', 

In [5]:
av_key_handle = open("alphavantage.apikey", "r")
ts = TimeSeries(key=av_key_handle.read().strip(), output_format='pandas')
av_key_handle.close()

In [13]:
dataframes = dict()
value_errors = set()
other_errors = set()
for ticker in tqdm_notebook(past_pdufa_syms):
    try:
        df, meta = ts.get_daily(symbol=ticker, outputsize='full')
        dataframes[meta["2. Symbol"]] = df
    except ValueError:
        value_errors.add(ticker)
    except:
        other_errors.add(ticker)

In [15]:
print value_errors
print other_errors

set(['VION', 'PCYC', 'FURX', 'DNDN', 'AIS', 'GEVA', 'BPAX', 'INSV', 'ZSPH', 'ADLR', 'CBRX', 'PPDI', 'ISIS', 'CYPB', 'HGSI', 'AEGR', 'ARIA', 'THRX', 'MDVN', 'HPTX', 'APPA', 'SLXP', 'SNTS', 'AVNR', 'AUXL', 'TSPT', 'QCOR', 'FLML', 'NEOL', 'ZGEN', 'DSCO', 'XNPT', 'SVNT', 'ALXA', 'FRX', 'NGSX', 'ISTA', 'NPSP', 'CLDA', 'RPTP', 'DYAX', 'CHTP', 'ITMN', 'RLYP', 'BIOD', 'ANAC', 'DRTX', 'KYTH', 'MELA', 'ANX', 'POZN'])
set(['ASTX', 'WCRX'])


In [39]:
dill.dump(dataframes, open('final_raw_dataframe_dict.pkl', 'w'))

###### Mini Checkpoint for slow API calls

In [40]:
dataframes = dill.load(open('final_raw_dataframe_dict.pkl', 'r'))

Now we'll run through our past FDA dates and join the FDA actions to each dataframe

In [41]:
company_list = dataframes.keys()

In [384]:
price_and_fda = dict()
for company in tqdm_notebook(company_list):
    company_events = []
    for event in FdaCal.events:
        matches = re.findall(tickerRe, event.name)
        if len(matches)>=1:
            if company in matches[0]:
                company_events.append((event.begin.datetime.strftime("%Y-%m-%d"), True))
    price = dataframes[company]
    raw_dates = pd.DataFrame(company_events, columns = ["date", "pdufa?"])
    dates = raw_dates.set_index("date")
    final = price.join(dates,rsuffix='_y')
    final['pdufa?'].fillna(value=False, inplace = True)
    price_and_fda[company] = final

That leaves us with a dict of dataframes containing every company's stock price, and FDA action dates

In [43]:
price_and_fda['ENTA'].head(3)

volume  close   high   open    low  pdufa?
2013-03-21  1763600.0  17.18  17.85  14.51  14.31   False
2013-03-22    75200.0  16.81  17.57  17.57  16.71   False
2013-03-25    24100.0  16.83  17.40  17.40  16.80   False

Now that I've got a good crop of downloaded data, lets cache it for good measure. 

In [187]:
dill.dump(price_and_fda, open("final_Prices_and_PDUFAs.pkl", "w"))

### Checkpoint 1 - FDA Action Dates Joined to Equity Prices

In [188]:
price_and_fda = dill.load(open("final_Prices_and_PDUFAs.pkl", "r"))

So thats every company's stock prices, with PDUFA dates going back to around 2006, and pricing data going back to 2001. More than enough data for our analysis. 

Lets unify all the prices into one frame, to construct a pharmaceutical price index. This will give us a base price to normalize our stock prices against, insulating our model from general economic events (the '08 housing crash) or events affecting the whole pharmaceutical sector (passage of new FDA regulations). 

In [46]:
price_and_fda = dill.load(open("Prices_and_PDUFAs_final", "r"))

In [47]:
first = True
for ticker, comp_df in price_and_fda.iteritems():
    if first:
        market_df = comp_df.copy()
        market_df.columns = ["volume-"+ticker,
                             "close-"+ticker,
                             "high-"+ticker,
                             "open-"+ticker,
                             "low-"+ticker,
                             "pdufa?-"+ticker]
        first = False
    else:
        market_df = pd.merge(market_df, comp_df, how='outer', left_index=True, right_index=True, suffixes=('', '-'+ticker))

In [48]:
price_mean = market_df.filter(regex='close').mean(axis = 1, skipna = True)
price_stdv = market_df.filter(regex='close').std(axis = 1, skipna = True)

In [49]:
stats_df = pd.merge(price_mean.to_frame(),
                    price_stdv.to_frame(), 
                    left_index=True, 
                    right_index=True, 
                    how='inner')
stats_df.rename(columns={u'0_x':"CP_mean", u'0_y':"CP_stdv"}, inplace=True)

In [50]:
stats_df.head()

CP_mean    CP_stdv
2000-01-03  28.502039  29.517285
2000-01-04  27.066072  28.246109
2000-01-05  27.154608  28.343677
2000-01-06  27.534455  28.559617
2000-01-07  29.768633  31.529028

This is as good a place as any to cache the closing price index

In [51]:
dill.dump(stats_df, open("close_price_stats_frame_final.pkl", "w"))

### Checkpoint 2

In [52]:
stats_df = dill.load(open("close_price_stats_frame_final.pkl", "r"))

Now I have the mean and standard deviation of close prices (`stats_df`) for every day of my data coverage. This will make it easy to normalize prices for every slice of time relevant to an FDA trial. 

Time to cut time slices for each clinical trial and generate a population of clinical trials and normalized prices.

In [53]:
norm_data = []
for company in tqdm_notebook(company_list):
    df = price_and_fda[company].join(stats_df, how='left').reset_index()
    pdufa_dates = df.index[df['pdufa?']].tolist()
    if len(pdufa_dates) > 0:
        for date in pdufa_dates:
            pRange = range(date-120, date-7)
            pCloses, pVolumes = [], []
            for i in pRange:
                try:
                    close_price = df.loc[i]['close']
                    volume = df.loc[i]['volume']
                    mean_price = df.loc[i]['CP_mean']
                    stdv_price = df.loc[i]['CP_stdv']
                    pCloses.append(( df.loc[i]['index'],(close_price-mean_price)/(stdv_price) ))
                    pVolumes.append(( df.loc[i]['index'], volume ))
                except:
                    pCloses.append(None)
                    pVolumes.append(None)
            norm_data.append((company, df.loc[date]['index'], (pCloses, pVolumes)))

Well we have normalized slices, lets add the annotations from our score sheet

In [54]:
scores = [line.split() for line in open("score_sheet_complete.txt", "r").readlines()]

In [55]:
norm_data_annotated = []
mismatches = []
for datum in tqdm_notebook(norm_data):
    for score in scores:
        if datum[0] == score [0] and datum [1] == score[1]:
            norm_data_annotated.append((datum[0], datum[1], score[2], datum[2] ))
            break

In [56]:
dill.dump(norm_data_annotated, open("normalized_training_data.pkl", "w"))

### Checkpoint 3

In [57]:
norm_data_annotated = dill.load(open("normalized_training_data.pkl", "r"))

Now we have normalized stock prices, in 120-7 day slices prior to FDA action dates. Lets pull those back into smaller pandas frames for feature extraction. 

In [58]:
def assemble_frame(datum):
    df = pd.DataFrame(datum[3][0], columns=['date','norm_price'])
    df['event'] = datum[0]+"/"+datum[1]
    df['outcome'] = int(datum[2])
    return df

In [59]:
first = True

for line in tqdm_notebook(norm_data_annotated):
    try:
        if first:
            agg_data = assemble_frame(line)
            first = False
        else:
            tmp_data = assemble_frame(line)
            agg_data = pd.concat([agg_data, tmp_data],ignore_index=True)
    except:
        print line[0], line[1], "failed"

COLL 2015-10-12 failed
NEOS 2015-11-09 failed



In [60]:
agg_data['date_stamp'] = pd.to_datetime(agg_data['date'])
event_labels = pd.factorize(agg_data['event'])
agg_data["event_stamp"] = event_labels[0]

Now lets remove out the trials will null prices on some days (either due to acquisitions or bankruptcies). 

In [61]:
agg_data['null'] = pd.isnull(agg_data).apply(lambda x: sum(x) , axis=1)
cleaned_agg = agg_data[agg_data['null'] == 0]

In [62]:
cleaned_agg.head()

date  norm_price            event  outcome date_stamp  event_stamp  \
0  2015-12-08   -0.133607  AAAP/2016-06-01        1 2015-12-08            0   
1  2015-12-09   -0.128218  AAAP/2016-06-01        1 2015-12-09            0   
2  2015-12-10   -0.126670  AAAP/2016-06-01        1 2015-12-10            0   
3  2015-12-11   -0.119063  AAAP/2016-06-01        1 2015-12-11            0   
4  2015-12-14   -0.110840  AAAP/2016-06-01        1 2015-12-14            0   

   null  
0     0  
1     0  
2     0  
3     0  
4     0

In [63]:
dill.dump(cleaned_agg, open('final_cleaned_price_slices.pkl', 'w'))

### Checkpoint 3 - Training data preprocessed

In [64]:
cleaned_agg = dill.load(open('final_cleaned_price_slices.pkl', 'r'))

That's a ready to extract package of every clinical trial scraped. Lets go ahead and make up a test and train split now, while its easy and convinent.

In [65]:
from sklearn.cross_validation import train_test_split

In [96]:
train_data, test_data = train_test_split(norm_data_annotated, train_size = .9)

In [97]:
first = True

for line in tqdm_notebook(train_data):
    try:
        if first:
            train_df = assemble_frame(line)
            first = False
        else:
            tmp_df = assemble_frame(line)
            train_df = pd.concat([train_df, tmp_df],ignore_index=True)
    except:
        print line[0], line[1], "failed"

train_df['date_stamp'] = pd.to_datetime(train_df['date'])
event_labels = pd.factorize(train_df['event'])
train_df["event_stamp"] = event_labels[0]

train_df['null'] = pd.isnull(train_df).apply(lambda x: sum(x) , axis=1)
train_clean = train_df[train_df['null'] == 0]

COLL 2015-10-12 failed
NEOS 2015-11-09 failed



In [98]:
first = True

for line in tqdm_notebook(test_data):
    try:
        if first:
            test_df = assemble_frame(line)
            first = False
        else:
            tmp_df = assemble_frame(line)
            test_df = pd.concat([test_df, tmp_df],ignore_index=True)
    except:
        print line[0], line[1], "failed"
test_df['date_stamp'] = pd.to_datetime(test_df['date'])
event_labels = pd.factorize(test_df['event'])
test_df["event_stamp"] = event_labels[0]

test_df['null'] = pd.isnull(test_df).apply(lambda x: sum(x) , axis=1)
test_clean = test_df[test_df['null'] == 0]

Thats two parts of a bifurcated dataframe. May as well cache it. 

In [99]:
dill.dump(train_clean, open("final_train_df.pkl", "w"))
dill.dump(test_clean, open("final_test_df.pkl", "w"))

### Checkpoint 4 - Test Train Split

In [100]:
train_clean = dill.load(open("final_train_df.pkl", "r"))
test_clean = dill.load(open("final_test_df.pkl", "r"))

Now for the serious work, extracting features from the pricing data in each case. 

I'll be using [tsfresh](http://tsfresh.readthedocs.io/en/latest/text/quick_start.html) to do the hard computing here, and then selecting the most relevant features. While I am able to compute almost 800 features for these data points, I'm going to narrow down to around ten of the most meaningful or important features. 

In [101]:
from tsfresh import extract_features

In [102]:
train_feats = extract_features(train_clean[['norm_price', 'event_stamp', 'date_stamp']], 
                              column_id="event_stamp", column_sort="date_stamp", 
                              column_value="norm_price", n_jobs=0).dropna(axis=1)

Feature Extraction: 100%|██████████| 209/209 [00:00<00:00, 25615.38it/s]


In [103]:
train_feats.head()

variable  norm_price__abs_energy  norm_price__absolute_sum_of_changes  \
id                                                                      
0                       8.267646                             0.632868   
1                       0.050117                             0.364996   
2                       6.148967                             1.010344   
3                     706.617635                             2.724921   
4                       4.240612                             0.662200   

variable  norm_price__agg_autocorrelation__f_agg_"mean"  \
id                                                        
0                                              0.338676   
1                                              0.216699   
2                                              0.110301   
3                                              0.062402   
4                                             -0.015751   

variable  norm_price__agg_autocorrelation__f_agg_"median"  \
id                                                          
0                                                0.291667   
1                                                0.061411   
2                                                0.130618   
3                                               -0.030932   
4                                               -0.165876   

variable  norm_price__agg_autocorrelation__f_agg_"var"  \
id                                                       
0                                             0.089200   
1                                             0.115088   
2                                             0.115312   
3                                             0.079144   
4                                             0.104243   

variable  norm_price__agg_linear_trend__f_agg_"max"__chunk_len_10__attr_"intercept"  \
id                                                                                    
0                                                 -0.193869                           
1                                                 -0.011717                           
2                                                 -0.219150                           
3                                                  2.515227                           
4                                                 -0.178728                           

variable  norm_price__agg_linear_trend__f_agg_"max"__chunk_len_10__attr_"rvalue"  \
id                                                                                 
0                                                 -0.807651                        
1                                                  0.330647                        
2                                                  0.246763                        
3                                                  0.432196                        
4                                                 -0.017051                        

variable  norm_price__agg_linear_trend__f_agg_"max"__chunk_len_10__attr_"slope"  \
id                                                                                
0                                                 -0.011369                       
1                                                  0.001689                       
2                                                  0.003207                       
3                                                  0.007583                       
4                                                 -0.000079                       

variable  norm_price__agg_linear_trend__f_agg_"max"__chunk_len_10__attr_"stderr"  \
id                                                                                 
0                                                  0.002625                        
1                                                  0.001524                        
2                                                  0.003983                        
3                                     

In [104]:
train_y =\
train_df[['event_stamp', 'outcome']]\
.groupby('event_stamp')\
.head(1).set_index('event_stamp')['outcome']

In [105]:
train_y.head()

event_stamp
0    0
1    1
2    0
3    1
4    0
Name: outcome, dtype: int64

In [106]:
test_feats = extract_features(test_clean[['norm_price', 'event_stamp', 'date_stamp']], 
                              column_id="event_stamp", column_sort="date_stamp", 
                              column_value="norm_price", n_jobs=0).dropna(axis=1)

Feature Extraction: 100%|██████████| 24/24 [00:00<00:00, 22201.87it/s]


In [107]:
test_feats.shape

(24, 622)

In [108]:
test_y =\
test_df[['event_stamp', 'outcome']]\
.groupby('event_stamp')\
.head(1).set_index('event_stamp')['outcome']

In [109]:
test_y.shape

(24,)

In [110]:
dill.dump(train_feats, open('final_train_features.pkl','w'))
dill.dump(test_feats, open('final_test_features.pkl','w'))

### Checkpoint 4 - Extracted Features

In [111]:
train_feats = dill.load(open("final_train_features.pkl", "r"))
test_feats = dill.load(open("final_test_features.pkl", "r"))

Now its time to pick out 10 or so meaningful features from the 622 possible features. Time for some reading. Then itll be time to apply those to a classification model. 

In [112]:
print"\n".join(list(train_feats.columns.values))

norm_price__abs_energy
norm_price__absolute_sum_of_changes
norm_price__agg_autocorrelation__f_agg_"mean"
norm_price__agg_autocorrelation__f_agg_"median"
norm_price__agg_autocorrelation__f_agg_"var"
norm_price__agg_linear_trend__f_agg_"max"__chunk_len_10__attr_"intercept"
norm_price__agg_linear_trend__f_agg_"max"__chunk_len_10__attr_"rvalue"
norm_price__agg_linear_trend__f_agg_"max"__chunk_len_10__attr_"slope"
norm_price__agg_linear_trend__f_agg_"max"__chunk_len_10__attr_"stderr"
norm_price__agg_linear_trend__f_agg_"max"__chunk_len_50__attr_"intercept"
norm_price__agg_linear_trend__f_agg_"max"__chunk_len_50__attr_"rvalue"
norm_price__agg_linear_trend__f_agg_"max"__chunk_len_50__attr_"slope"
norm_price__agg_linear_trend__f_agg_"max"__chunk_len_50__attr_"stderr"
norm_price__agg_linear_trend__f_agg_"max"__chunk_len_5__attr_"intercept"
norm_price__agg_linear_trend__f_agg_"max"__chunk_len_5__attr_"rvalue"
norm_price__agg_linear_trend__f_agg_"max"__chunk_len_5__attr_"slope"
norm_price__agg_li

In [159]:
features_of_interest = ['norm_price__mean',
                        'norm_price__median',
                        'norm_price__mean_change',
                        #'norm_price__mean_abs_change',
                        'norm_price__first_location_of_maximum',
                        'norm_price__first_location_of_minimum',
                        'norm_price__linear_trend__attr_"slope"',
                        'norm_price__count_above_mean',
                        'norm_price__count_below_mean'
                       ]

In [160]:
print train_feats[features_of_interest].shape
train_feats[features_of_interest].head()

(209, 8)


variable  norm_price__mean  norm_price__median  norm_price__mean_change  \
id                                                                        
0                -0.265791           -0.249245                -0.001468   
1                -0.010960           -0.008207                 0.000124   
2                -0.229526           -0.226107                -0.001035   
3                 2.499430            2.514329                 0.001829   
4                -0.192816           -0.190763                 0.000123   

variable  norm_price__first_location_of_maximum  \
id                                                
0                                      0.353982   
1                                      0.283186   
2                                      0.000000   
3                                      0.805310   
4                                      0.946903   

variable  norm_price__first_location_of_minimum  \
id                                                
0                                      0.991150   
1                                      0.097345   
2                                      0.380531   
3                                      0.008850   
4                                      0.725664   

variable  norm_price__linear_trend__attr_"slope"  \
id                                                 
0                                      -0.001221   
1                                       0.000295   
2                                       0.000399   
3                                       0.001168   
4                                      -0.000093   

variable  norm_price__count_above_mean  norm_price__count_below_mean  
id                                                                    
0                                 79.0                          34.0  
1                                 70.0                          43.0  
2                                 60.0                          53.0  
3                                 66.0                          47.0  
4                                 58.0                          55.0

In [161]:
print test_feats[features_of_interest].shape
test_feats[features_of_interest].head()

(24, 8)


variable  norm_price__mean  norm_price__median  norm_price__mean_change  \
id                                                                        
0                -0.421011           -0.409471                -0.000908   
1                -0.231476           -0.232116                -0.000483   
2                -0.557929           -0.558836                 0.000011   
3                -0.251067           -0.240050                 0.000064   
4                -0.284424           -0.289434                -0.000335   

variable  norm_price__first_location_of_maximum  \
id                                                
0                                      0.557522   
1                                      0.539823   
2                                      0.955752   
3                                      0.123894   
4                                      0.318584   

variable  norm_price__first_location_of_minimum  \
id                                                
0                                      0.929204   
1                                      0.256637   
2                                      0.628319   
3                                      0.699115   
4                                      0.867257   

variable  norm_price__linear_trend__attr_"slope"  \
id                                                 
0                                      -0.000749   
1                                      -0.000024   
2                                      -0.000100   
3                                      -0.000282   
4                                      -0.000814   

variable  norm_price__count_above_mean  norm_price__count_below_mean  
id                                                                    
0                                 81.0                          32.0  
1                                 56.0                          57.0  
2                                 51.0                          62.0  
3                                 82.0                          31.0  
4                                 54.0                          59.0

Thats our split data, with our features of interest. Lets begin Modeling. 

In [166]:
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV, cross_val_score
from sklearn.metrics import accuracy_score

In [178]:
scaler = StandardScaler()
classifier = SVC(C=1, coef0=1, degree=1)
params = {"C":range(1,5),
          "degree":range(1,3),
          "coef0":range(1,3)
         }
classifier_gs = GridSearchCV(classifier, params)

In [179]:
classifier_gs.fit(scaler.fit_transform(train_feats), train_y)

GridSearchCV(cv=None, error_score='raise',
       estimator=SVC(C=1, cache_size=200, class_weight=None, coef0=1,
  decision_function_shape='ovr', degree=1, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'C': [1, 2, 3, 4], 'coef0': [1, 2], 'degree': [1, 2]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score=True,
       scoring=None, verbose=0)

In [180]:
classifier_gs.best_params_

{'C': 1, 'coef0': 1, 'degree': 1}

In [181]:
cross_val_score(classifier, scaler.transform(test_feats), y=test_y)

array([ 0.66666667,  0.75      ,  0.71428571])

Thats a trained and cross validated model. Lets pickle it for safe keeping.

In [185]:
dill.dump(classifier, open("final_trained_svc.pkl","w"))

### Checkpoint 5 - Trained Model

In [186]:
classifier = dill.load(open("final_trained_svc.pkl","r"))

Now that we have a working predictor, lets play with some visualizations to show how powerful is can be.

To start, lets collect our data into one large frame to explore (actually, run the same processing steps on the non-split data).

In [189]:
all_feats = extract_features(cleaned_agg[['norm_price', 'event_stamp', 'date_stamp']], 
                              column_id="event_stamp", column_sort="date_stamp", 
                              column_value="norm_price", n_jobs=0).dropna(axis=1)

Feature Extraction: 100%|██████████| 233/233 [00:00<00:00, 30255.19it/s]


In [193]:
cleaned_agg

date  norm_price            event  outcome date_stamp  \
0      2015-12-08   -0.133607  AAAP/2016-06-01        1 2015-12-08   
1      2015-12-09   -0.128218  AAAP/2016-06-01        1 2015-12-09   
2      2015-12-10   -0.126670  AAAP/2016-06-01        1 2015-12-10   
3      2015-12-11   -0.119063  AAAP/2016-06-01        1 2015-12-11   
4      2015-12-14   -0.110840  AAAP/2016-06-01        1 2015-12-14   
5      2015-12-15   -0.124347  AAAP/2016-06-01        1 2015-12-15   
6      2015-12-16   -0.143909  AAAP/2016-06-01        1 2015-12-16   
7      2015-12-17   -0.135699  AAAP/2016-06-01        1 2015-12-17   
8      2015-12-18   -0.093418  AAAP/2016-06-01        1 2015-12-18   
9      2015-12-21   -0.132192  AAAP/2016-06-01        1 2015-12-21   
10     2015-12-22   -0.132920  AAAP/2016-06-01        1 2015-12-22   
11     2015-12-23   -0.139685  AAAP/2016-06-01        1 2015-12-23   
12     2015-12-24   -0.139910  AAAP/2016-06-01        1 2015-12-24   
13     2015-12-28   -0.141184  AAAP/2016-06-01        1 2015-12-28   
14     2015-12-29   -0.144858  AAAP/2016-06-01        1 2015-12-29   
15     2015-12-30   -0.136939  AAAP/2016-06-01        1 2015-12-30   
16     2015-12-31   -0.120907  AAAP/2016-06-01        1 2015-12-31   
17     2016-01-04   -0.107698  AAAP/2016-06-01        1 2016-01-04   
18     2016-01-05   -0.111013  AAAP/2016-06-01        1 2016-01-05   
19     2016-01-06   -0.111977  AAAP/2016-06-01        1 2016-01-06   
20     2016-01-07   -0.100521  AAAP/2016-06-01        1 2016-01-07   
21     2016-01-08   -0.109069  AAAP/2016-06-01        1 2016-01-08   
22     2016-01-11   -0.119822  AAAP/2016-06-01        1 2016-01-11   
23     2016-01-12   -0.166242  AAAP/2016-06-01        1 2016-01-12   
24     2016-01-13   -0.107799  AAAP/2016-06-01        1 2016-01-13   
25     2016-01-14   -0.138169  AAAP/2016-06-01        1 2016-01-14   
26     2016-01-15   -0.135811  AAAP/2016-06-01        1 2016-01-15   
27     2016-01-19   -0.100612  AAAP/2016-06-01        1 2016-01-19   
28     2016-01-20   -0.108369  AAAP/2016-06-01        1 2016-01-20   
29     2016-01-21   -0.087918  AAAP/2016-06-01        1 2016-01-21   
...           ...         ...              ...      ...        ...   
26299  2016-02-03   -0.433812   OPK/2016-03-29        1 2016-02-03   
26300  2016-02-04   -0.433603   OPK/2016-03-29        1 2016-02-04   
26301  2016-02-05   -0.424206   OPK/2016-03-29        1 2016-02-05   
26302  2016-02-08   -0.429827   OPK/2016-03-29        1 2016-02-08   
26303  2016-02-09   -0.426723   OPK/2016-03-29        1 2016-02-09   
26304  2016-02-10   -0.427178   OPK/2016-03-29        1 2016-02-10   
26305  2016-02-11   -0.422916   OPK/2016-03-29        1 2016-02-11   
26306  2016-02-12   -0.438558   OPK/2016-03-29        1 2016-02-12   
26307  2016-02-16   -0.441752   OPK/2016-03-29        1 2016-02-16   
26308  2016-02-17   -0.434540   OPK/2016-03-29        1 2016-02-17   
26309  2016-02-18   -0.431808   OPK/2016-03-29        1 2016-02-18   
26310  2016-02-19   -0.431634   OPK/2016-03-29        1 2016-02-19   
26311  2016-02-22   -0.416595   OPK/2016-03-29        1 2016-02-22   
26312  2016-02-23   -0.417552   OPK/2016-03-29        1 2016-02-23   
26313  2016-02-24   -0.417946   OPK/2016-03-29        1 2016-02-24   
26314  2016-02-25   -0.410862   OPK/2016-03-29        1 2016-02-25   
26315  2016-02-26   -0.408995   OPK/2016-03-29        1 2016-02-26   
26316  2016-02-29   -0.410956   OPK/2016-03-29        1 2016-02-29   
26317  2016-03-01   -0.401830   OPK/2016-03-29        1 2016-03-01   
26318  2016-03-02   -0.412393   OPK/2016-03-29        1 2016-03-02   
26319  2016-03-03   -0.402579   OPK/2016-03-29        1 2016-03-03   
26320  2016-03-04   -0.400589   OPK/2016-03-29        1 2016-03-04   
26321  2016-03-07   -0.397767   OPK/2016-03-29        1 2016-03-07   
26322  2016-03-08   -0.404487   OPK/2016-03-29        1 2016-03-08   
26323  2016-03-09   -0.398363   OPK/2016-03-29        1 2016-03-09   
26324  2016-03-10   -0.

In [190]:
print all_feats[features_of_interest].shape
all_feats[features_of_interest].head()

(233, 8)


variable  norm_price__mean  norm_price__median  norm_price__mean_change  \
id                                                                        
0                -0.002417            0.035755                 0.000851   
1                 0.049569            0.051360                -0.000588   
2                 0.095391            0.076302                 0.001551   
3                 0.053320            0.049479                 0.000757   
4                 0.041048            0.035759                -0.000833   

variable  norm_price__first_location_of_maximum  \
id                                                
0                                      0.610619   
1                                      0.681416   
2                                      0.893805   
3                                      0.699115   
4                                      0.000000   

variable  norm_price__first_location_of_minimum  \
id                                                
0                                      0.203540   
1                                      0.973451   
2                                      0.681416   
3                                      0.336283   
4                                      0.530973   

variable  norm_price__linear_trend__attr_"slope"  \
id                                                 
0                                       0.002417   
1                                       0.000558   
2                                       0.000700   
3                                       0.001458   
4                                      -0.000554   

variable  norm_price__count_above_mean  norm_price__count_below_mean  
id                                                                    
0                                 60.0                          53.0  
1                                 60.0                          53.0  
2                                 36.0                          77.0  
3                                 55.0                          58.0  
4                                 52.0                          61.0

In [191]:
all_y =\
cleaned_agg[['event_stamp', 'outcome']]\
.groupby('event_stamp')\
.head(1).set_index('event_stamp')['outcome']

In [198]:
all_events =\
cleaned_agg[['event_stamp','event']]\
.groupby('event_stamp')\
.head(1).set_index('event_stamp')['event']

In [322]:
all_predictions = classifier_gs.predict(scaler.transform(all_feats))

In [332]:
events_and_predictions = pd.DataFrame(all_events).join(pd.DataFrame(all_predictions))

In [333]:
events_and_predictions.shape

(233, 2)

Lets generate a random population of guesses to check, In this case, I'm going to generate random 1,000,000 guesses (of 1 or 0) for each row of the data and average these to generate a random distribution. 

In [334]:
random_guesses = np.random.randint(0,2,size=(events_and_predictions.shape[0], 1000000))

In [335]:
print random_guesses.shape
random_guesses

(233, 1000000)


array([[1, 1, 0, ..., 0, 1, 1],
       [0, 0, 0, ..., 1, 1, 1],
       [0, 0, 0, ..., 1, 0, 0],
       ..., 
       [0, 1, 0, ..., 1, 0, 1],
       [1, 1, 0, ..., 1, 1, 0],
       [1, 1, 0, ..., 0, 0, 0]])

In [336]:
random_guess_means = np.mean(random_guesses, axis = 1)

In [337]:
random_guess_means.shape

(233,)

In [338]:
events_and_predictions.columns = ['event', 'pass_prediction']
events_and_predictions['pass_random'] = [int(x) for x in random_guess_means.round()]
events_and_predictions['pass_all'] = 1

In [339]:
events_and_predictions

event  pass_prediction  pass_random  pass_all
event_stamp                                                         
0            AAAP/2016-06-01                1            1         1
1            AAAP/2016-12-28                1            0         1
2            AAAP/2017-08-28                1            1         1
3            BSTC/2009-02-27                1            1         1
4            BSTC/2009-08-28                0            0         1
5            BSTC/2013-12-06                1            1         1
6            BSTC/2014-10-20                1            1         1
7            ELTP/2016-07-14                0            0         1
8             GSK/2013-12-18                0            1         1
9             GSK/2014-01-08                0            0         1
10            GSK/2014-01-09                0            0         1
11            GSK/2014-04-15                0            1         1
12            GSK/2015-11-04                1            0         1
13           CLVS/2016-06-28                0            0         1
14           CLVS/2017-02-23                1            1         1
15           IRWD/2012-08-30                0            0         1
16           VNDA/2014-01-31                0            1         1
17           VNDA/2016-05-27                1            1         1
18           ACUR/2009-06-30                1            0         1
19           ACUR/2011-06-17                0            0         1
20           DEPO/2013-05-31                0            1         1
21           CHMA/2016-04-15                0            1         1
22           VRTX/2011-05-23                1            1         1
23           VRTX/2014-03-27                0            1         1
24           VRTX/2015-03-17                1            0         1
25           VRTX/2016-09-30                1            1         1
26           OREX/2010-12-07                1            1         1
27           OREX/2011-01-31                0            0         1
28           OREX/2014-09-11                0            1         1
29           PCRX/2011-10-28                0            1         1
...                      ...              ...          ...       ...
203          CELG/2012-10-12                1            0         1
204          CELG/2017-02-24                1            1         1
205          CELG/2017-08-30                1            1         1
206          PDLI/2012-08-10                0            1         1
207          PDLI/2013-07-01                0            0         1
208          AKRX/2008-06-02                0            0         1
209          HOLX/2010-09-24                0            1         1
210          UTHR/2009-04-30                0            0         1
211          UTHR/2012-10-23                1            0         1
212          UTHR/2013-03-25                1            1         1
213          AEZS/2014-11-05                0            1         1
214          RDHL/2014-02-03                1            1         1
215          ALXN/2011-10-07                1            0         1
216          ALXN/2015-12-08                1            0         1
217          TSRO/2017-01-11                1            1         1
218          TSRO/2017-06-30                0            1         1
219          IPXL/2015-01-09                1            0         1
220           PFE/2015-04-13                1            1         1
221          ALKS/2010-09-16                1            1         1
222          ALKS/2010-10-12                1            0         1
223          ALKS/2010-10-22                1            0         1
224          ALKS/2017-06-05                1            0         1
225          IPCI/2017-09-25                0            1         1
226          BDSI/2015-10-23                0            0         1
227           PDL/2012-08-10                0            0         1
228           PDL/201

In [340]:
predicted_passes = events_and_predictions[events_and_predictions['pass_prediction'] == 1]
predicted_fails = events_and_predictions[events_and_predictions['pass_prediction'] == 0]

Thats every event in a human readable format, with it's machine predicted outcome, a randomly guessed outcome, and an all outcomes guess.
Now lets pull some price data and see how each model fares.

In [354]:
price_and_fda['AAAP'].loc[u'2016-06-01']

volume      73059
close       29.46
high      30.3876
open         29.8
low        28.811
pdufa?       True
Name: 2016-06-01, dtype: object

In [425]:
def x_days_later(date_str, x):
    pdufa_day = datetime.strptime(date_str,"%Y-%m-%d")
    change = timedelta(days = x)
    delta_date = pdufa_day + change
    return delta_date.strftime("%Y-%m-%d")

In [426]:
x_days_later('2016-06-01', -7)

'2016-05-25'

In [462]:
prior_and_post_prices = []
for stamp in events_and_predictions['event']:
    ticker, date = stamp.split("/")
    try:
        p_7_day = price_and_fda[ticker].loc[x_days_later(date, -7)]['close']
    except KeyError:
        p_7_day = None
    try:
        p_60_day = price_and_fda[ticker].loc[x_days_later(date,60)]['close']
    except KeyError:
        try:
            p_60_day = price_and_fda[ticker].loc[x_days_later(date,59)]['close']
        except KeyError:
            try:
                p_60_day = price_and_fda[ticker].loc[x_days_later(date,58)]['close']
            except KeyError:
                try:
                    p_60_day = price_and_fda[ticker].loc[x_days_later(date,57)]['close']
                except KeyError:
                    p_60_day = None
    line = (stamp, p_7_day, p_60_day)
    if None not in line:
        prior_and_post_prices.append(line)

In [465]:
prior_and_post_prices = pd.DataFrame(prior_and_post_prices)
prior_and_post_prices.columns = ['event', 'close_-7_Day', 'close_+60_Day']
prior_and_post_prices

event  close_-7_Day  close_+60_Day
0    AAAP/2016-06-01       29.2900        31.2900
1    AAAP/2016-12-28       24.6800        36.9600
2    AAAP/2017-08-28       48.6500        72.9100
3    BSTC/2009-02-27       19.9600        19.5000
4    BSTC/2009-08-28       26.2100        30.3600
5    BSTC/2013-12-06       22.1900        22.2000
6    BSTC/2014-10-20       34.5000        39.5100
7    ELTP/2016-07-14        0.3518         0.1855
8     GSK/2013-12-18       51.4300        55.8900
9     GSK/2014-01-09       52.7900        55.8400
10    GSK/2014-04-15       52.1900        54.5000
11    GSK/2015-11-04       43.5300        40.3500
12   CLVS/2016-06-28       13.1800        23.1500
13   CLVS/2017-02-23       61.3700        56.9200
14   IRWD/2012-08-30       12.1500        12.1100
15   VNDA/2014-01-31       12.9000        16.8800
16   VNDA/2016-05-27        9.5950        11.2000
17   ACUR/2009-06-30        5.8900         6.1100
18   ACUR/2011-06-17        3.8200         3.1100
19   DEPO/2013-05-31        5.8900         6.3800
20   CHMA/2016-04-15       12.7200         2.5700
21   VRTX/2011-05-23       57.2700        50.9800
22   VRTX/2014-03-27       77.7700        71.0400
23   VRTX/2015-03-17      124.1400       126.2100
24   VRTX/2016-09-30       90.6300        85.6500
25   OREX/2010-12-07        5.1900         3.6050
26   OREX/2011-01-31        8.9900         2.8800
27   OREX/2014-09-11        5.9900         5.1300
28   PCRX/2011-10-28        9.6500         8.8000
29   PCRX/2015-03-05      115.9900        66.8700
..               ...           ...            ...
192  MDCO/2015-06-23       28.7900        30.9400
193  MDCO/2017-08-29       35.4100        28.3000
194  CELG/2011-06-17       58.3000        56.1800
195  CELG/2012-10-12       80.2900        82.0800
196  CELG/2017-02-24      121.1600       125.5200
197  CELG/2017-08-30      129.1900        98.1700
198  PDLI/2012-08-10        6.9100         8.0800
199  PDLI/2013-07-01        7.4100         7.9400
200  HOLX/2010-09-24       16.5400        16.4400
201  UTHR/2009-04-30       56.6500        83.5200
202  UTHR/2012-10-23       56.5900        52.5100
203  UTHR/2013-03-25       61.8700        66.3500
204  AEZS/2014-11-05        1.1100         0.6001
205  RDHL/2014-02-03       12.6500        14.4900
206  ALXN/2011-10-07       64.0600        65.0500
207  ALXN/2015-12-08      180.2800       142.2800
208  TSRO/2017-01-11      138.3200       173.3000
209  TSRO/2017-06-30      146.7200       124.2300
210  IPXL/2015-01-09       31.8000        38.4600
211   PFE/2015-04-13       34.4800        34.2100
212  ALKS/2010-09-16       13.7200        11.1100
213  ALKS/2010-10-12       14.9200        11.2700
214  ALKS/2010-10-22       14.8000        12.5700
215  BDSI/2015-10-23        6.8000         4.7500
216   PDL/2012-08-10        4.4500         0.0000
217   PDL/2013-07-01        0.0000         0.0000
218   PDL/2014-11-19        6.4470         6.2910
219  TTNP/2016-05-27        6.3400         5.0000
220   EBS/2016-08-15       29.7700        29.2400
221   OPK/2016-03-29       11.6000        10.4200

[222 rows x 3 columns]

In [467]:
predictions_and_prices =\
pd.merge(events_and_predictions, prior_and_post_prices, on='event')